### ISSUE TITLE
[Bug] Android FPS drops to unusable levels until the app freezes

### ISSUE THREAD

**Issue by dxt298** [Unity-Technologies/arfoundation-samples#768]
**Describe the bug**
A developer of ours has our app freeze after it runs for a while (shortest time taken: 2min, average of around 5min or so).
The "freeze" seems to be the app constantly slowing down (low fps) until it is straight unusable.

Thinking that it might be our app, and to attempt to narrow down what's going on, I created a Test Scene in which essentially only the AR Session and AR Camera exist.  But despite there being nothing in the scene (no tracking managers are active either), the app still ends up freezing for him (though, it does take longer, like 10minutes or more).

I have personally tested our app (and that test scene) on the following devices with no issues:
iPad 6th Gen
iPad 2020
Samsung Galaxy S8

I originally thought it had to do with frame rate and whatnot, so I removed setting Application.frameRate, which fixed the choppy issues that occurred on both of the android devices, but did not solve the eventual freezing on the tablet.

Attached is the profiler data with GPU profiling on -- for that empty test scene
[profilerdata3.data.zip](https://github.com/Unity-Technologies/arfoundation-samples/files/6173756/profilerdata3.data.zip)

At a loss for what to do, especially since it eventually freezes up on an empty scene with just AR basics in it.

**Smartphone (please complete the following information):**
 - Device: Samsung Tab S6 Lite
 - OS: Android 10/11 (Was originally on 10, then his tablet updated to 11, and the issue still occurs)
 - Unity version: 2020.2.2f1
 - ARFoundation version: 4.0.12

**Comment by tdmowrer** [comment]
We haven't gotten any other reports like this, so I wonder if it is specific to that device. To rule out AR Foundation as a culprit, could you try one of the scenes from [Google's ARCore SDK for Unity](https://github.com/google-ar/arcore-unity-sdk)?

Please also submit a bug here: https://unity3d.com/unity/qa/bug-reporting

**Comment by dxt298** [comment]
Hey there, thanks for the response.

Our dev with that device has been super busy working on stuff so we havent gotten around to specifically building the google arcore project (there's apparently a lot of errors to deal with too when downloading it fresh).

On the other hand, I took a look at the issues in the google arcore repo and I noticed some with descriptions somewhat similar to our problem, so it may not be an AR Foundation issue, but I'll update again when we get the chance to actually try out the other project.

**Comment by dxt298** [comment]
Update: 
- Got the google arcore project to run for him
- The issue occurs even then

I'll try to see if I can get a smaller profiler data file (the one he has is 500mb+), but for now here are some screenshots

![image](https://user-images.githubusercontent.com/7322280/113049799-b2113d80-9169-11eb-9931-032905f6209d.png)
![image (1)](https://user-images.githubusercontent.com/7322280/113049812-b3db0100-9169-11eb-90c8-5efab6578c7f.png)

**Comment by tdmowrer** [comment]
> * Got the google arcore project to run for him
> * The issue occurs even then

Have you created a GitHub issue on [their page](https://github.com/google-ar/arcore-unity-sdk/issues)?

From the screenshots, it looks like it might just be trying to draw a very large mesh on a slower-than-average device. Hard to tell, though.

**Comment by dxt298** [comment]
Yeah, so he can make the "freeze" occur faster in the AR Core project by clicking to place objects (this is the HelloAR scene)
Those two screenshots are from that sort of situation.

I wasn't too concerned at first when he told me our app slows down because we do load a pretty big item, but when he told me it still freezes up with no tracking managers going on, then I started to get worried.

As for an issue on their page, I actually read into one of them that seemed similar to this https://github.com/google-ar/arcore-unity-sdk/issues/773

I actually think this may be it?  I asked him if his device supported depth api, and he told me that the popup said it did not.

We're not having very good luck trying to get an older version of the plugin to install on his device so we can test it (getting weird build/install errors), but I with the depth API thing in mind, I tried it out on a Galaxy S7 we have lying around.

I wasn't able to install 1.15 like the issue mentioned, instead going with 1.17
I can't tell if it's the phone being such a potato or whatnot, but the camera essentially stops working like a minute or two into the scene.  I am still able to press on any planes that exist in the phone's screen to place an object though.  Turning off the point cloud renderer allows that minute or two to be a bit longer but it'll still end up the same.

If I'm correct, 1.17 shouldn't have depth API so if the theory was on point, this should run ok on the S7 but it doesn't seem to.

Will report back whenever we actually attempt to get it working on his device.

**Comment by dxt298** [comment]
Updates:

- The "FPS drops to unusable levels" is limited to the Tab S6 Lite
- On an S7, the camera feed freezes, but the app still runs (can see an object's animation going on, can place objects onto planes, etc, but the camera feed is frozen)
- The assumption of the depth API being the culprit makes no sense since 4.0.12 uses AR Core 1.17, which is before that is implemented (but it just so happens the devices where this occurs doesnt have depth API support according to google's list
- Have actually managed to 'solve' the issue (no slowdowns, no camera feed freezes, etc) by switching from mono to il2cpp, and targeting only arm64
- As for why/how, I am not sure myself.

**Comment by tdmowrer** [comment]
The change from 32 bit to 64 bit stands out to me as the most likely culprit here. If you are releasing an app to the Play Store, you should target both anyway.

> The assumption of the depth API being the culprit makes no sense since 4.0.12 uses AR Core 1.17

Where did you get this information? The ARCore XR Plugin package version 4.0.12 uses 1.19 of the ARCore SDK (it was updated in 4.0.9; see [changelog](https://docs.unity3d.com/Packages/com.unity.xr.arcore@4.0/changelog/CHANGELOG.html#409---2020-10-06)).

**Comment by shenrique38** [comment]
> Updates:
> 
> * The "FPS drops to unusable levels" is limited to the Tab S6 Lite
> * On an S7, the camera feed freezes, but the app still runs (can see an object's animation going on, can place objects onto planes, etc, but the camera feed is frozen)
> * The assumption of the depth API being the culprit makes no sense since 4.0.12 uses AR Core 1.17, which is before that is implemented (but it just so happens the devices where this occurs doesnt have depth API support according to google's list
> * Have actually managed to 'solve' the issue (no slowdowns, no camera feed freezes, etc) by switching from mono to il2cpp, and targeting only arm64
> * As for why/how, I am not sure myself.

I also faced the same issue, and the proposed solution described above worked for me to solve it. In my case I was trying to run an empty scene in the android tablet Samsung Galaxy Tab S3 9.7 SM-T820 32GB, built from unity 2021.1.5f1. After 2 or 5 min of running the empty secene, the constant 30fps would drop to 0 and make the app to freeze and stop working. The same issue was not perceived in another tablet, Samsung Galaxy Tab S6 10.5 SM-T860 128GB, using the same app built. After following the changes described in the post, the app now works in teh Tab S3 without problems.

**Comment by Yang192** [comment]
Same thing happens to me on Samsung Galaxy A32 5G and Redmi Note 7. The game freezes in about 2 to 5 minutes, and sometimes returns to normal after a few minutes. How can I solve this problem?

OS : Android 11 , Unity version: 2021.3.6f1, ARFoundation version: 4.2.3

**Comment by Yang192** [comment]
I disabled all my scripts in the game, and only kept the script that prints the FPS and time of the game. NativeHeap always goes up from 38MB to 590MB after 370~380 seconds and causes the game to freeze. After 10~65 seconds, NativeHeap will drop to a value that the game can run well. What should I set to make NativeHeap stable?
![1659042616546](https://user-images.githubusercontent.com/70422638/181640270-abc70669-2a61-4a80-a4ff-578b6129316b.jpg)
![66383](https://user-images.githubusercontent.com/70422638/181640282-df165b42-fefd-44f4-b3ee-5dc2d795db78.jpg)
